myia/grad.py

```diff
@@ -0,0 +1,284 @@
+        self.sensitivity_nodes[(graph.output, graph)] = bparam
+
+        self.todo_fw.add(graph)
+        self.todo_bw.add(graph)
+
+        return Constant(tgraph), Constant(bgraph)
+
+    def process_graph(self, graph):
+        """Process this graph and return the forward graph.
+        
+        This may not work if called more than once with different
+        graphs. Create a new Grad instance instead.
+        """
+        self.process_all_graphs_forward(graph)
+        self.process_all_graphs_backward()
+        return self.tagged_graphs[graph]
+
+    def process_graph_forward(self, graph):
+        """Create the forward graph."""
+        if graph in self.done_fw:
+            return
+
+        if graph not in self.tagged_graphs:
+            self.scaffold_graph(graph)
+
+        tgraph = self.tagged_graphs[graph]
+        bgraph = self.backpropagator_graphs[graph]
+
+        # Return (↑graph.output, ♢graph). The first element is given
+        # by the `phi` method.
+        tgraph.output = Apply([
+            mktuple,
+            self.phi(graph.output),
+            Constant(bgraph)
+        ], tgraph)
+
+        self.done_fw.add(graph)
+
+    def process_all_graphs_forward(self, root):
+        """Create the forward graph for all graphs starting from root."""
+        self.todo_fw.add(root)
+        while self.todo_fw:
+            g = self.todo_fw.pop()
+            self.process_graph_forward(g)
+
+    def process_graph_backward(self, graph):
+        """Create the backward graph."""
+        bgraph = self.backpropagator_graphs[graph]
+
+        # Return ((∇fv1, ∇fv2, ...), ∇arg1, ∇arg2, ...)
+        # Where ∇x is given by `rho(x, graph)`
+        bgraph.output = Apply([
+            mktuple,
+            Apply([
+                mktuple,
+                *[self.rho(p, graph)
+                  for p in self.fv_order[graph]]
+            ], bgraph),
+            *[self.rho(p, graph)
+              for p in graph.parameters]
+        ], bgraph)
+
+    def process_all_graphs_backward(self):
+        """Create the backward graph for all graphs.
+
+        This method has no argument: it simply processes all graphs for which we
+        created a forward graph.
+        """
+        for g in self.todo_bw:
+            self.process_graph_backward(g)
+
+    def phi(self, node):
+        """Compute equivalent node in forward graph."""
+        if node in self.tagged_nodes:
+            return self.tagged_nodes[node]
+
+        tg = node.graph and self.tagged_graphs[node.graph]
+
+        if is_graph(node):
+            # We will have to process this graph too.
+            tagged, bprop = self.scaffold_graph(node.value)
+            self.graph_to_ct[node.value].add(node)
+        elif is_constant(node):
+            # Note that this application will have its graph set to None, which
+            # makes sense since it's basically a constant expression.
+            tagged, bprop = Apply([J, node], tg), None
```

Comment:
I'm not sure about changing the semantics this way. It seems to complicate things like VMs and scheduling i.e. when do you actually evaluate these expressions? And when do they go out of scope? Ideally we should just "evaluate" them at compile time, or otherwise we can wrap the entire graph in a global graph whose only job it is to evaluate these nodes and then start the rest of the calculations.

Reply:
Well, there was already the idea before of wrapping all graphs in a global graph which would be fed the global variables. That seems like an appropriate scope for all these computations.